model.py

```diff
@@ -1606,20 +1622,60 @@ def decode_two_part(self, _db, username, password):
         """Decode a short client token that has already been split into
         two parts.
         """
-        try:
-            signature = self.adobe_base64_decode(password)
-        except Exception, e:
-            raise ValueError("Invalid password: %s" % password)
-        return self._decode(_db, username, signature)
+        library = patron_identifier = account_id = None
+
+        # No matter how we do this, if we're going to create
+        # a DelegatedPatronIdentifier, we need to extract the Library
+        # and the library's identifier for this patron from the 'username'
+        # part of the token.
+        library, expires, patron_identifier = self._split_token(
+            _db, username
+        )
 
-    def _decode(self, _db, token, supposed_signature):
-        """Make sure a client token is properly formatted, correctly signed,
-        and not expired.
+        # First see if a delegate can give us an Adobe ID (account_id)
+        # for this patron.
+        for delegate in self.delegates:
+            try:
+                account_id, label, content = delegate.sign_in_standard(
+                    username, password
+                )
+            except Exception, e:
+                # This delegate couldn't help us.
+                pass
+            if account_id:
+                # We got it -- no need to keep checking delegates.
+                break
+
+        if not account_id:
+            # The delegates couldn't help us; let's try to do it
+            # ourselves.
+            try:
+                signature = self.adobe_base64_decode(password)
+            except Exception, e:
+                raise ValueError("Invalid password: %s" % password)
+
+            patron_identifier, account_id = self._decode(
+                _db, username, signature
+            )
```

Comment:
Maybe this could return the library and expires so you don't have to call split_token twice.

Reply:
I experimented with this idea and it makes the code a lot more complicated for a pretty small performance boost. This is because that first split_token call serves two purposes: if the username isn't actually a Short Client Token, split_token raises an exception which lets us bail out immediately.

It's possible (in fact, likely) that a delegated server could give a response to a username/password that's not a Short Client Token. Without the bailout early in `decode_two_part`, we might send an HTTP request to a foreign server, trigger its side effects, and then discover that we jumped the gun. In that case we'll need to bail out anyway and have the calling code make the HTTP request again.

It's possible there's some design that's altogether different and solves both these problems, but the inefficiency is small enough that I'd prefer to leave it alone.